compiler/rustc_lint/src/non_local_def.rs

```diff
@@ -0,0 +1,243 @@
+use std::convert::identity;
+
+use rustc_hir::{
+    Body, FnRetTy, GenericArg, GenericBound, GenericParam, GenericParamKind, Item, ItemKind, Path,
+    QPath, TyKind, WherePredicate,
+};
+use rustc_span::{def_id::DefId, sym, MacroKind};
+
+use crate::{lints::NonLocalDefinitionsDiag, LateContext, LateLintPass, LintContext};
+
+declare_lint! {
+    /// The `non_local_definitions` lint checks for `impl` blocks and `#[macro_export]`
+    /// macro inside bodies (functions, enum discriminant, ...).
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// trait MyTrait {}
+    /// struct MyStruct;
+    ///
+    /// fn foo() {
+    ///     impl MyTrait for MyStruct {}
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Creating non local definitions go against expectation and can create decrepencies
+    /// in tooling. In should be avoided.
+    pub NON_LOCAL_DEFINITIONS,
+    Warn,
+    "checks for non local definitions"
+}
+
+#[derive(Default)]
+pub struct NonLocalDefinitions {
+    is_in_body: u32,
+}
+
+impl_lint_pass!(NonLocalDefinitions => [NON_LOCAL_DEFINITIONS]);
+
+impl<'tcx> LateLintPass<'tcx> for NonLocalDefinitions {
+    fn check_body(&mut self, _cx: &LateContext<'tcx>, _body: &'tcx Body<'tcx>) {
+        self.is_in_body += 1;
+    }
+
+    fn check_body_post(&mut self, _cx: &LateContext<'tcx>, _body: &'tcx Body<'tcx>) {
+        self.is_in_body -= 1;
+    }
+
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx Item<'tcx>) {
+        if self.is_in_body > 0 {
+            match item.kind {
+                ItemKind::Impl(impl_) => {
+                    // The RFC states:
+                    //
+                    // > An item nested inside an expression-containing item (through any
+                    // > level of nesting) may not define an impl Trait for Type unless
+                    // > either the **Trait** or the **Type** is also nested inside the
+                    // > same expression-containing item.
+                    //
+                    // To achieve this we get try to get the paths of the _Trait_ and
+                    // _Type_, and we look inside thoses paths to try a find in one
+                    // of them a type whose parent is the same as the impl definition.
+                    //
+                    // If that's the case this means that this impl block decleration
+                    // is using local items and so we don't lint on it.
+
+                    let parent_impl = cx.tcx.parent(item.owner_id.def_id.into());
+
+                    let generics_has_local_parent =
+                        generics_params_has_local_parent(&impl_.generics.params, cx, parent_impl)
+                            || generics_predicates_has_local_parent(
+                                &impl_.generics.predicates,
+                                cx,
+                                parent_impl,
+                            );
```

Comment:
Doesn't this allow
```rust
trait Trait {}
struct Struct;

fn bad() {
    struct Local;
    impl Trait for Struct where Local: Any {}
}
```
?

Reply:
Yes, it was. I've now removed any predicates checks.